### PR TITLE
Math: FFT: Add 16 bit FFT and IFFT version

### PR DESCRIPTION
**Pull request by singalsu**
First commit switches to script generated twiddle_32.h

Second commit adds the 16 bit FFT version.

The FFT API has changed with added number argument 16 or 32 to fft_plan_new() to indicate desired word length. Fft_execute() has been replaced with fft_execute_16() and fft_execute_32() functions. Please let me know what you think about API change. I'm not aware of other FFT library usage than cmocka.

### CODE REVIEW

File: test/cmocka/src/math/fft/fft.c [hunk 1]
```diff
@@ -54,7 +54,7 @@ static void fft_real(struct comp_buffer *src, struct comp_buffer *dst, uint32_t
 	if (!outb)
```
Comment (singalsu) [inline test/cmocka/src/math/fft/fft.c:54]: Need to add cmocka test case for 16 bit FFT and IFFT. Note also that FFT cmocka tests are not built and executed. Need to address that also.

**Comment by lgirdwood** [comment]
@ShriramShastry @andrula-song any comments ?

**Review by lgirdwood — COMMENTED** [review]
Minor comments from me.

File: src/include/sof/math/fft.h [hunk 2]
```diff
@@ -75,9 +82,59 @@ static inline void icomplex_shift(struct icomplex32 *input, int32_t n, struct ic
...
 }
 
+/*
+ * Helpers for 16 bit FFT calculation
+ */
+
+static inline void icomplex16_add(struct icomplex16 *in1, struct icomplex16 *in2,
+				  struct icomplex16 *out)
+{
+	out->real = in1->real + in2->real;
```
[the remark this answers is not in the document]
Comment (singalsu) [reply src/include/sof/math/fft.h:92]: Yep, xtensa version will follow this quarter.

File: test/cmocka/src/math/fft/fft.c [hunk 3]
```diff
@@ -204,106 +220,195 @@ static void fft_real_2(struct comp_buffer *src, struct comp_buffer *dst1,
...
 	printf("%s: peak at point %d\n", __func__, r);
 
 	/* the peak should be in range i +/-1 */
 	assert_in_range(r, i - 1, i + 1);
+
+	/* the min. SNR should be met */
+	noise = integrate_power_32(out, 0, i - 2);
+	signal = integrate_power_32(out, i - 1, i + 1);
+	noise += integrate_power_32(out, i + 2, fft_size / 2 - 1);
+	snr = 10 * log10(signal / noise);
```
[the remark this answers is not in the document]
Comment (singalsu) [reply test/cmocka/src/math/fft/fft.c:344]: I already noticed some some and fixed. Got much lower SNR than expected,  bugs fixed it reached 130 dB SNR for 32 bit 256 size and pretty good 120 dB for 1024 also. I get the same result on native x86 and xt-run test.

Divide with integer is a caveat. E.g. a / 48000 vs. a / 48000.0. But in this case with cmocka test it's most convenient and least error prone to use double or float math for evaluating performance and create test input. Doing tests with fractional fixed math would create another potential error source.

[the remark this answers is not in the document]
Comment (lgirdwood) [reply test/cmocka/src/math/fft/fft.c:344]: type casting integers to floats/doubles will be fine (but not other way round).

File: src/include/sof/math/fft.h [hunk 4]
```diff
@@ -21,12 +21,19 @@ struct icomplex32 {
 	int32_t imag;
 };
 
+struct icomplex16 {
+	int16_t real;
+	int16_t imag;
+};
```
[the remark this answers is not in the document]
Comment (singalsu) [reply src/include/sof/math/fft.h:27]: @lgirdwood Wow, I  thought this kind of structure is packed by default. Build with __attribute__((packed)) made 2.1 times the MCPS. I suppose this means that 16 bit version is not saving at all RAM with complex number storage if I omit the packing? Is there a way to help compiler to vectorize with packing included?

[the remark this answers is not in the document]
Comment (singalsu) [reply src/include/sof/math/fft.h:27]: Packed also increases processing time with gcc but the effect is not as much, 1.4x only. But it's best to keep it non-packed for now and have different data structure definition for xtensa intrinsics.

[the remark this answers is not in the document]
Comment (lgirdwood) [reply src/include/sof/math/fft.h:27]: Lets capture this in a comment next to the structure. We may have to reformat the structure differently for Hifi3,4 and for SSE, AVX etc.

File: src/math/fft/fft.c [hunk 5]
```diff
@@ -5,14 +5,21 @@
...
 
-#include <sof/audio/coefficients/fft/twiddle_32.h>
 #include <sof/audio/buffer.h>
 #include <sof/audio/format.h>
 #include <sof/common.h>
 #include <sof/lib/alloc.h>
 #include <sof/math/fft.h>
 
-struct fft_plan *fft_plan_new(struct icomplex32 *inb, struct icomplex32 *outb, uint32_t size)
+#if CONFIG_MATH_16BIT_FFT
```
Comment (singalsu) [inline src/math/fft/fft.c:14]: I think I need to split fft.c into separate 16 bit and 32 bit modules and use cmakelist.txt to select based on kconfig the sources. This allows cmocka test to build e.g. 32 bit library and test it even if platform configuration has disabled it. I ran into problem with auditory library. It depends on FFT and it won't build if CONFIG_MATH_xxBIT_FFT are not defined. When there are no #if CONFIGs in module they can be always built to unit test.

Comment (lgirdwood) [reply src/math/fft/fft.c:14]: Splitting is good, as it allows Kconfig to select for size vs quality.

**Comment by lgirdwood** [comment]
@singalsu can you force push again to restart CI now that power is back.

**Review by lgirdwood — APPROVED** [review]

**Pull request merged**